File: 14. Longest Common Subsequence (LCS).cpp
// 14. Longest Common Subsequence (LCS)
// Topic: Longest Common Subsequence
// Description: Given strings a and b, find length of LCS using recursion + memo.
// Input: a b (two strings on separate lines or same line)
// Output: length of LCS

#include <bits/stdc++.h>
using namespace std;

int lcs(int i, int j, const string& a, const string& b, vector<vector<int>>& memo)
{
    if (i >= (int)a.size() || j >= (int)b.size())
        return 0;
    if (memo[i][j] != -1)
        return memo[i][j];
    if (a[i] == b[j])
        return memo[i][j] = 1 + lcs(i + 1, j + 1, a, b, memo);
    return memo[i][j] = max(lcs(i + 1, j, a, b, memo), lcs(i, j + 1, a, b, memo));
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    string a, b;
    if (!getline(cin, a))
        return 0;
    if (!getline(cin, b))
        return 0;
    vector<vector<int>> memo(a.size(), vector<int>(b.size(), -1));
    cout << lcs(0, 0, a, b, memo) << '\n';
}

// Memoization process started in lcs where memo[i][j] caches LCS length for suffixes starting at i
// and j. Time Complexity: O(|a| * |b|) Space Complexity: O(|a| * |b|)

File: 14. Longest Common Subsequence (LCS).py
# 14. Longest Common Subsequence (LCS)
# Topic: Longest Common Subsequence
# Description: Given strings a and b, find length of LCS using recursion + memoization.
# Input: two lines: string a and string b
# Output: length of LCS

import sys
from functools import lru_cache

if __name__ == '__main__':
    lines = sys.stdin.read().splitlines()
    if not lines:
        sys.exit(0)
    a = lines[0].rstrip('\n')
    b = lines[1].rstrip('\n') if len(lines) > 1 else ''
    n, m = len(a), len(b)

    @lru_cache(maxsize=None)
    def lcs(i: int, j: int) -> int:
        if i == n or j == m:
            return 0
        if a[i] == b[j]:
            return 1 + lcs(i+1, j+1)
        return max(lcs(i+1, j), lcs(i, j+1))

    print(lcs(0,0))

# Memoization process started at lcs decorated with lru_cache.
# Time Complexity: O(n * m)
# Space Complexity: O(n * m)